src/goto-instrument/ada_checks.cpp

```diff
@@ -0,0 +1,182 @@
+  {
+    Overflow,
+    Range,
+    DivByZero
+  };
+
+  void replace_by_option(
+    const exprt &condition,
+    goto_programt &goto_program,
+    goto_programt::targett instruction,
+    const std::string &function_name,
+    const std::string &replace_option);
+
+  Ada_Checkt get_check_type(const std::string &function_name) const
+  {
+    if(function_name.find("Overflow") != std::string::npos)
+      return Ada_Checkt::Overflow;
+    if(function_name.find("Range") != std::string::npos)
+      return Ada_Checkt::Range;
+    if(function_name.find("Division") != std::string::npos)
+      return Ada_Checkt::DivByZero;
+    throw invalid_source_file_exceptiont{"unsupported Ada check: " +
+                                         function_name};
+  }
+
+  Replace_Typet get_replace_type(const std::string &replace_option) const
+  {
+    if(replace_option == "assert-first-arg")
+      return Replace_Typet::AssertOnly;
+    if(replace_option == "assume-first-arg")
+      return Replace_Typet::AssumeOnly;
+    if(replace_option == "assert-then-assume-first-arg")
+      return Replace_Typet::AssertThenAssume;
+    throw invalid_command_line_argument_exceptiont{
+      "unknown Ada check replacement option" + replace_option, "--ada-checks"};
+  }
+
+  std::string type2string(Ada_Checkt type) const
+  {
+    switch(type)
+    {
+    case Ada_Checkt::Overflow:
+      return "Overflow";
+    case Ada_Checkt::Range:
+      return "Range";
+    case Ada_Checkt::DivByZero:
+      return "Division by zero";
+    }
+    UNREACHABLE;
+  }
+};
+
+void ada_checkst::replace_with(const std::string &replace_option)
+{
+  auto &goto_functions = goto_model.goto_functions;
+  for(auto &goto_function_pair : goto_functions.function_map)
+  {
+    goto_programt &function_body = goto_function_pair.second.body;
+    for(goto_programt::targett instruction = function_body.instructions.begin();
+        instruction != function_body.instructions.end();
+        ++instruction)
+    {
+      if(instruction->is_function_call())
+      {
+        const auto &function_call = instruction->get_function_call();
+        const auto &function = function_call.function();
+
+        const auto &function_name =
+          id2string(to_symbol_expr(function).get_identifier());
+        if(has_prefix(function_name, CPROVER_ADA_PREFIX))
+        {
+          const auto &function_arguments = function_call.arguments();
+          CHECK_RETURN(!function_arguments.empty());
+
+          replace_by_option(
+            *function_arguments.begin(),
+            function_body,
+            instruction,
+            function_name,
+            replace_option);
+        }
+      }
+    }
+  }
+}
+
+void ada_checkst::replace_by_option(
+  const exprt &condition,
+  goto_programt &goto_program,
+  goto_programt::targett instruction,
+  const std::string &function_name,
+  const std::string &replace_option)
+{
+  exprt boolean_condition = condition;
+  while(boolean_condition.id() == ID_typecast)
```

Comment:
Why is this `while`? I thought the `to_typecast_expr` below always simplifies? Is it possible it returns something that has the same structure, and hasn't been structurally reduced?

Reply:
Okay, ignore this comment. After I wrote it, I remembered that the `to_typecast_expr` is not even simplifying, it's just a cast. But I still wonder why the need for `while` here.

Reply:
In a C program this is probably unnecessary, but gnat2goto can produce nested typecasts.